### PR TITLE
Rework config

### PR DESCRIPTION
**Pull request by jensdrenhaus**
The separation between host config and security config is not relevant to the internal state of the stboot command. The idea is to introduce a general Opts struct meant to store all configuration and settings for stboot. Still, the parsing of separated host and security settings from remain untouched.
The source of host config and security config settings may vary, e.g. JSON, cmd line, hard drive, NVRAM, ...

### CODE REVIEW

**Comment by bjoto** [comment]
This is a nice cleanup! Some thoughts:

* Why is the configuration package called stboot? Isn't that a bit weird?
* Are you planning to integrate the package into stboot.go prior merging. I would be in favor of that.

**Comment by jensdrenhaus** [comment]
Naming is a thing. And I'm not done with thinking about it yet. 
I aim to have on struct that holds every setting for stboot to operate
```
type Opts struct {
// ...
}
```
Inside a package `config` it would be used as `config.Opts` which is wired. So, I came up with `stboot.Opts`
package `stboot` could hold further common things like consts and vars with default values, etc. 

Maybe the validation and parsing code will go in another pkg. 
And sure, once ready, it will be integrated in `stboot.go` still work in progress ;)

**Comment by bjoto** [comment]
Gotcha. TBH it was *now* that I realized that this was a draft, and not a PR. Ugh. ;-)

**Comment by sylv-io** [comment]
fyi: creating a `stboot` directory breaks the Go/build job. Either we have to rename the build output file or the directory.

**Comment by jensdrenhaus** [comment]
changed the name from stboot to opts. Still WIP. 
SecurityCfg is done. HostConfig is to be done.

**Comment by jensdrenhaus** [comment]
This major rework improves the parsing and validation of host_config and security_config. Package `config` is being completely replaced by `opts` which serves as single point for stboot configuration. 

ATM, I tried to touch `stboot.go` as little as possible - just enough to bring the API of opts in. This enables us to start discussing of a proper flag handling regarding source of host_config now. You are welcome to write your proposals. This can be done here or with a follow-up PR?!

Further enhancements to be done in `opts`:
* Improve the validation part with cross validations (dependencies `Security`<->`HostCfg`
* Add more integration tests
* Create an API for the validation mechanism, e.g. to be used while creating host_config.json with `stmgr`
* Pull tings like HTTPS and signing root certificates, timestamp file, etc. into `opts.Opts`

File: opts/hostcfg.go [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+		dnsServer = h.DNSServer.String()
+	}
+	if h.NetworkInterface != nil {
+		networkInterface = h.NetworkInterface.String()
+	}
+	for _, u := range h.ProvisioningURLs {
+		provisioningURLs = append(provisioningURLs, u.String())
+	}
+
+	easy := struct {
```
Comment (sylv-io) [inline opts/hostcfg.go:98]: why `easy` as var name? :smile:

File: stboot.go [hunk 2]
```diff
@@ -235,25 +206,25 @@ func main() {
...
-		case config.DynamicIP:
-			if err := network.ConfigureDHCP(hostConfig, *doDebug); err != nil {
+		case opts.IPDynamic:
+			if err := network.ConfigureDHCP(&stOptions.HostCfg, *doDebug); err != nil {
 				stlog.Error("cannot set up IO: %v", err)
 				host.Recover()
 			}
 		default:
-			stlog.Error("unknown network mode: %s", hostConfig.IPAddrMode.String())
+			stlog.Error("unknown network mode: %s", stOptions.IPAddrMode.String())
```
Comment (sylv-io) [inline stboot.go:222]: same here, no default required.

File: opts/opts.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+type InvalidError string
+
+func (e InvalidError) Error() string {
+	return string(e)
+}
+
+type validator func(*Opts) error
+
+// Loader wraps the Load funktion.
```
Comment (sylv-io) [inline opts/opts.go:13]: function

**Review by sylv-io — CHANGES_REQUESTED** [review]

**Comment by sylv-io** [comment]
Regarding functionality, it looks like there are no regressions :+1:. This branch works on the system-transparency testing workflow too:
https://github.com/system-transparency/system-transparency/actions/runs/1722663592

File: opts/security.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (s *Security) UnmarshalJSON(data []byte) error {
+	var maybeCfg map[string]interface{}
+	if err := json.Unmarshal(data, &maybeCfg); err != nil {
+		return err
+	}
+
+	// check for missing json tags
+	tags, _ := jsonTags(s)
```
Comment (MDr164) [inline opts/security.go:72]: Unchecked error?

File: opts/security_test.go [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+			}
+			if string(got) != tt.want {
+				t.Errorf("got %q, want %q", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestSecurityUnmarshalJSON(t *testing.T) {
+
```
Comment (MDr164) [inline opts/security_test.go:84]: Blank line here?

File: opts/validation.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
+package opts
+
+import "strings"
+
+var (
```
Comment (MDr164) [inline opts/validation.go:5]: Not sure if it makes sense here but none of those exported errors are commented.

File: opts/validation.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	}
+	return nil
+}
+
+func hasAllowdChars(s string) bool {
+	if len(s) > 64 {
+		return false
+	}
+	for _, c := range s {
+		if (c < 'a' || c > 'z') && (c < 'A' || c > 'Z') && (c < '0' || c > '9') && c != '-' && c != '_' {
```
Comment (MDr164) [inline opts/validation.go:121]: Doesn't using the regex std package make more sense here?

File: opts/validation_test.go [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+			if err != tt.want {
+				t.Errorf("got %q, want %q", err, tt.want)
+			}
+
+		})
+	}
+}
+
+func TestHostCfgValidation(t *testing.T) {
+
```
Comment (MDr164) [inline opts/validation_test.go:84]: Blank line here?

**Review by MDr164 — CHANGES_REQUESTED** [review]
nit: Several JSON files have no trailing newline while some others do. GitHub complains about those that don't have them. Was that intended to be inconsistent there? It shouldn't make any difference for the test cases I guess.

**Comment by sylv-io** [comment]
> nit: Several JSON files have no trailing newline while some others do. GitHub complains about those that don't have them. Was that intended to be inconsistent there? It shouldn't make any difference for the test cases I guess.

Run this command to add all missing new lines:
```bash
for f in $(git ls-files| grep -v vmlinuz); do sed -i '$a\' $f; done
```

File: opts/opts.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+// Opts controlls the operation of stboot
+type Opts struct {
+	Version int
+	Security
+	HostCfg
+}
+
+// NewOpts return a new Opts initialized by the provided Loaders
+func NewOpts(loaders ...Loader) (*Opts, error) {
+	opts := &Opts{Version: 0}
```
[the remark this answers is not in the document]
Comment (sylv-io) [reply opts/opts.go:28]: btw is this even necessary? The zero value of Version (int) is zero :smile:

File: stboot.go [hunk 10]
```diff
@@ -641,50 +612,53 @@ func diskLoad(names []string) ([]*ospkgSampl, error) {
...
-		// STDATA network cache dir
-		p = filepath.Join(host.DataPartitionMountPoint, host.NetworkOSpkgCache)
-		stat, err := os.Stat(p)
-		if err != nil || !stat.IsDir() {
-			return fmt.Errorf("STDATA: missing directory %s", host.NetworkOSpkgCache)
-		}
-	}
-	return nil
-}
+// NEEDS REWORK regarding STDATA
```
[the remark this answers is not in the document]
Comment (jensdrenhaus) [reply stboot.go:615]: STDATA?

File: stboot.go [hunk 11]
```diff
@@ -147,71 +129,60 @@ func main() {
 			stlog.Error("mount STDATA partition: %v", err)
 			host.Recover()
 		}
-		if err = validatePartitions(securityConfig.BootMode); err != nil {
```
[the remark this answers is not in the document]
Comment (jensdrenhaus) [reply stboot.go:150]: Good point. I remove the validation() function completely for now. This needs to be done together with the STDATA rework

File: stboot.go [hunk 12]
```diff
@@ -286,7 +257,7 @@ func main() {
 		}
 		ospkgSampls = append(ospkgSampls, ss...)
 	default:
-		stlog.Error("unsupported boot mode: %s", securityConfig.BootMode.String())
+		stlog.Error("unsupported boot mode: %s", stOptions.BootMode.String())
```
[the remark this answers is not in the document]
Comment (jensdrenhaus) [reply stboot.go:260]: Right. However, I would still go with a default to catch that invalid state, but I redo the message

File: opts/hostcfg.go (hunk 1 above)
Comment (jensdrenhaus) [reply opts/hostcfg.go:98]: easy to marshal ;)

File: stboot.go (hunk 2 above)
Comment (jensdrenhaus) [reply stboot.go:222]: Right. However, I would still go with a default to catch that invalid state, but I redo the message

File: opts/opts.go (hunk 3 above)
Comment (jensdrenhaus) [reply opts/opts.go:13]: thx

File: opts/hostcfg_test.go [hunk 13]
```diff
@@ -0,0 +1,561 @@
...
+			//remove all white space
+			want := strings.Join(strings.Fields(tt.want), "")
+			if string(got) != want {
+				t.Errorf("got %s, want %s", got, want)
+			}
+		})
+	}
+}
+
+const (
```
[the remark this answers is not in the document]
Comment (jensdrenhaus) [reply opts/hostcfg_test.go:170]: Left from previous iteration. Clean it up

File: opts/integration_test.go [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+	"encoding/json"
+	"net/url"
+	"os"
+	"path/filepath"
+	"reflect"
+	"testing"
+)
+
+func TestOptsLoader(t *testing.T) {
+	tmp, err := os.MkdirTemp("", "")
```
[the remark this answers is not in the document]
Comment (jensdrenhaus) [reply opts/integration_test.go:13]: I was not aware of that one. Nice!

File: opts/hostcfg.go [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+			Value: fmt.Sprintf("string %q", s),
+			Type:  reflect.TypeOf(i),
+		}
+	}
+	*i = ip
+	return nil
+}
+
+// SecurityCfg groups host specific configuration.
+type HostCfg struct {
```
[the remark this answers is not in the document]
Comment (jensdrenhaus) [reply opts/hostcfg.go:61]: Yes, later on. I like to reworkextend the overall docu and see how it looks like at https://pkg.go.dev/github.com/system-transparency/stboot

File: opts/opts.go [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+type validator func(*Opts) error
+
+// Loader wraps the Load funktion.
+// Load fills particular fields of Opts depending on its source.
+type Loader interface {
+	Load(*Opts) error
+}
+
+// Opts controlls the operation of stboot
+type Opts struct {
```
[the remark this answers is not in the document]
Comment (jensdrenhaus) [reply opts/opts.go:20]: Yes, later on with the docu rework. 
ATM just make sure, there is some minimal meaningful comment.

File: opts/opts.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jensdrenhaus) [reply opts/opts.go:28]: Yeah.. we need to think how and if we will use Version anyway.

File: opts/security.go [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+)
+
+// BootMode controlls where to load the OS from.
+type BootMode int
+
+const (
```
[the remark this answers is not in the document]
Comment (jensdrenhaus) [reply opts/security.go:16]: Taken from the type definition

File: opts/security.go [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+			Value: fmt.Sprintf("string %q", s),
+			Type:  reflect.TypeOf(b),
+		}
+	}
+	*b = bm
+	return nil
+}
+
+// Security groups security configuration.
+type Security struct {
```
[the remark this answers is not in the document]
Comment (jensdrenhaus) [reply opts/security.go:58]: Taken form struct comment

File: opts/security.go (hunk 4 above)
Comment (jensdrenhaus) [reply opts/security.go:72]: Reworked the function

File: opts/validation.go (hunk 6 above)
Comment (jensdrenhaus) [reply opts/validation.go:5]: Makes no sense :D ?!

**Comment by jensdrenhaus** [comment]
> ```shell
> for f in $(git ls-files| grep -v vmlinuz); do sed -i '$a\' $f; done
> ```

Done

File: opts/hostcfg.go [hunk 19]
```diff
@@ -0,0 +1,285 @@
...
+// UnmarshalJSON implements json.Unmarshaler.
+func (i *IPAddrMode) UnmarshalJSON(data []byte) error {
+	toId := map[string]IPAddrMode{
+		"":       IPUnset,
+		"static": IPStatic,
+		"dhcp":   IPDynamic,
+	}
+
+	var s string
+	err := json.Unmarshal(data, &s)
```
Comment (MDr164) [inline opts/hostcfg.go:44]: if err := json.Unmarshal....

File: opts/integration_test.go [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+		IPAddrMode:       IPDynamic,
+		ProvisioningURLs: []*url.URL{provURL},
+	}
+	hcPath := filepath.Join(tmp, "hostconfig.json")
+
+	scb, err := json.Marshal(&sc)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = os.WriteFile(scPath, scb, os.ModePerm)
```
Comment (MDr164) [inline opts/integration_test.go:33]: if err := os.WriteFile...

File: opts/integration_test.go [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	hcb, err := json.Marshal(&hc)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = os.WriteFile(hcPath, hcb, os.ModePerm)
```
Comment (MDr164) [inline opts/integration_test.go:41]: see above

**Review by MDr164 — APPROVED** [review]

File: opts/hostcfg.go (hunk 1 above)
Comment (sylv-io) [reply opts/hostcfg.go:98]: haha

**Review by sylv-io — APPROVED** [review]

**Pull request merged**